tests/fixtures/ddp_fixtures.py

```diff
@@ -1,70 +1,55 @@
+from typing import Callable
 
-import _pytest.config
-import _pytest.config.argparsing
-import _pytest.fixtures
-import _pytest.mark
 import pytest
-
-_MAIN_PYTEST_KEY = "MAIN_PYTEST"
-
-
-@pytest.fixture
-def is_main_pytest_process() -> bool:
-    if _MAIN_PYTEST_KEY in os.environ:
-        return False
-    else:
-        os.environ[_MAIN_PYTEST_KEY] = "1"
-        return True
+import torch.distributed
 
 
 @pytest.fixture(autouse=True)
-def reset_osenviron(is_main_pytest_process: None):
-    # Require the is_main_pytest_process fixture so that will be set before the env variable capture
-    original_environ = os.environ.copy()
-    try:
-        yield
-    finally:
-        # remove the extra keys
-        keys_to_remove: List[str] = []
-        for k in os.environ.keys():
-            if k not in original_environ:
-                keys_to_remove.append(k)
-        for k in keys_to_remove:
-            del os.environ[k]
-
-        # reset the other keys
-        for k, v in original_environ.items():
-            os.environ[k] = v
-
-
-@pytest.fixture(autouse=True)
-def ddp_fork_multi_pytest(request: _pytest.fixtures.FixtureRequest, reset_osenviron: None) -> None:
-    # When using DDP fork, the same command is executed again.
-    # When multiple tests are executed with a single pytest command,
-    # then the DDP fork will re-run all tests, when instead we just want
-    # the forked test to run again. To prevent this behavior, we store the
-    # running pytest id as an environ. If this variable is set, and it does
-    # NOT match the current pytest test ID, then the test should be skipped
-    # as the test was launched by a different DDP fork
-    test_name: str = request.node.name
-    if "PYTEST_TEST_ID" not in os.environ:
-        os.environ["PYTEST_TEST_ID"] = test_name
-        return
-    if os.environ["PYTEST_TEST_ID"] != test_name:
-        pytest.skip("Skipping test invoked within a DDP fork")
-
-
-@pytest.fixture
-def ddp_tmpdir(tmpdir: pathlib.Path) -> str:
-    if os.environ.get("DDP_TMP_DIR") is None:
-        # Need to set the tmpdir as an environ, so all subprocesses share the same tmpdir.
-        # Upon subprocess calls, pytest gives each ddp process a different tmpdir
-        # Instead, we want to use the same tmpdir for test case assertions.
-        os.environ['DDP_TMP_DIR'] = str(tmpdir)
-        return str(tmpdir)
-    return os.environ["DDP_TMP_DIR"]
+def ddp_cleanup():
+    yield
+    if torch.distributed.is_initialized():
+        torch.distributed.destroy_process_group()
+
+
+def _dist_run_target(target: Callable, rank: int, num_procs: int, *args, **kwargs):
+    os.environ['WORLD_SIZE'] = str(num_procs)
+    os.environ['RANK'] = str(rank)
+    os.environ['LOCAL_RANK'] = str(rank)
+    os.environ['LOCAL_WORLD_SIZE'] = str(num_procs)
+
+    os.environ['MASTER_ADDR'] = '127.0.0.1'
+    os.environ['MASTER_PORT'] = str(29400)
+
+    target(*args, **kwargs)
+
+
+def with_distributed(num_procs: int,
```

Comment:
I love the simplicity of this....but also wonder what are the potential implication of using `mp.spawn` for tests while we run jobs with the launch script. Here are some questions that come to mind:

* Is this slow? (Especially after running many tests, and then we spawn something?)
* I vaguely remember that some things did not work with mp.spawn (maybe that was never an issue) but if so, curious how we fixed it?
* Do we have any testing for the launch script?
* Are there any issues with using a TCPstore (i.e. port already in use?)
* Does this work with wandb? Mainly thinking if we end up doing long-running regressions with pytest, then we would be using spawn.
* Curious about the function signature -- could it take the targets's `*args` and `**kwargs`, invoke `run_target`,  return None? I think that is more standard than returning a callable.

As an alternative, thoughts about invoking pytest using the launch script? (E.g. `composer -m pytest ....`. Also see the comment in the launch script for the `-m` option). All the CPU tests would execute once per rank (not a big deal). For the 1 GPU tests, we would need to make the 2nd process idle for the duration of the test. It would also be good to have some sort of coordination to make sure that tests start at the same time across all ranks (to ensure we don't hit a timeout stemming from a slow cleanup of a previous test).

Reply:
Great questions - here are my thoughts:

- I seem to have broken the tests after rebasing on latest `dev`, but I'll report benchmarks for `pytest` once that's fixed.
- `spawn` does admittedly have a notable issue - it depends on being able to pickle the local state. This wasn't a serious blocker for any of the tests I refactored, but I do see how it could become an issue.
- There's no testing for the launch script at the moment - I'll add a test :)
- Honestly I expected to run into issues with TCPstore reuse, but for some reason that never happened. If that does become an issue down the road somehow, this function can handle some of the logic of searching for an available port.
- I don't see any reason this wouldn't work with WandB. The means of multiprocessing isn't too relevant to logging.
- I think I stand by the function signature. The main advantage it has is it makes it easy to differentiate between arguments for `with_distributed` (especially those optional timeout arguments), and the arguments for the target function.

Reply:
`run_long` takes 110s for me before this change; 140s after.

Reply:
Hmm kk...30s isn't a killer, but would be concerned if this delta grows with O(# of ddp tests).

This is fine for now, but would like to experiment in a future PR of doing `composer -m pytest ...`.

Reply:
Yeah, that makes sense to me. I'm hoping at least that the need for N > 2 in multiprocessing tests should be extremely limited.